src/exo/cursors.py

```diff
@@ -0,0 +1,591 @@
+
+    def after(self, dist=1) -> Gap:
+        attr, _range = self._path[-1]
+        assert len(_range) > 0
+        return self.parent().child(attr, _range.stop - 1).after(dist)
+
+    def prev(self, dist=1) -> Cursor:
+        # TODO: what should this mean?
+        #  1. The node after the block?
+        #  2. The block shifted over?
+        #  3. The block of nodes leading up to the start?
+        raise NotImplementedError('Block.prev')
+
+    def next(self, dist=1) -> Cursor:
+        # TODO: what should this mean?
+        #  1. The node after the block?
+        #  2. The block shifted over?
+        #  3. The block of nodes past the end?
+        raise NotImplementedError('Block.next')
+
+    # ------------------------------------------------------------------------ #
+    # Sequence interface implementation
+    # ------------------------------------------------------------------------ #
+
+    def __iter__(self):
+        attr, _range = self._path[-1]
+        block = self.parent()
+        for i in _range:
+            yield block.child(attr, i)
+
+    def __getitem__(self, i):
+        attr, r = self._path[-1]
+        r = r[i]
+        if isinstance(r, range):
+            if r.step != 1:
+                raise IndexError('block cursors must be contiguous')
+            return Block(self._proc, self._path[:-1] + [(attr, r)])
+        else:
+            return self.parent().child(attr, r)
+
+    def __len__(self):
+        _, _range = self._path[-1]
+        return len(_range)
+
+    # ------------------------------------------------------------------------ #
+    # AST mutation
+    # ------------------------------------------------------------------------ #
+
+    def _replace(self, nodes: list, *, empty_default=None):
+        """
+        This is an UNSAFE internal function for replacing a block in an AST with
+        a list of statements and providing a forwarding function as collateral.
+        It is meant to be package-private, not class-private, so it may be
+        called from other internal classes and modules, but not from end-user
+        code.
+        """
+        assert self._path
+        assert len(self) > 0
+
+        def update(node):
+            attr, i = self._path[-1]
+            children = getattr(node, attr)
+            new_children = children[:i.start] + nodes + children[i.stop:]
+            new_children = new_children or empty_default or []
+            return node.update(**{attr: new_children})
+
+        from .API import Procedure
+        p = Procedure(self._rewrite_node(update))
+
+        return p, self._forward_replace(weakref.ref(p), len(nodes))
+
+    def _forward_replace(self, new_proc, n_ins):
+        del_range = self._path[-1][1]
+        n_diff = n_ins - len(del_range)
+
+        def fwd_node(i):
+            if i in del_range:
+                raise InvalidCursorError('node no longer exists')
+            return i + n_diff * (i >= del_range.stop)
```

Comment:
This is overly strong when the cursor points to an invariant subtree.  i.e. the assumption here is that we don't use `replace` to replace something like `s1;s2` with `s2;s1` or even `s` with `if c: s`.

Reply:
Yeah, we'll need something slightly different like `._wrap` for a block-wrapping replacement. Or maybe something smarter that I'm not thinking of? The semantics of `._replace` here are really meant to be "this is totally new code" like what happens if you use the `.replace(instr, block)` directive or if you do something dramatic in the simplifier, etc.

Reply:
I could imagine that this `._replace` ends up being a special case of `._wrap` where whatever tells the forwarding function where the old nodes went is `None` or something.

Reply:
Ahh, so here's a thought.  Suppose that your general replacement (or forwarding?) call doesn't just accept a cursor defining the context (i.e. `C[•]`) in which the rewriting is happening, but also accepts a set of cursors to the "invariant sub-trees" along with cursors/paths to their new locations?  (maybe there's a way to be slightly more concise?)  If that were the case, then we would do the current checks (is the cursor being forwarded pointed at some context higher up in the tree?  Then just forward it trivially) but also do checks w.r.t. sub-trees.  (i.e. if the cursor is pointing at some node within an invariant sub-tree, then modify the pre-sub-tree prefix of the path to account for the relocation of the sub-tree.

Reply:
Note that in this view of things, gaps "in sub-trees" only get forwarded if they're properly _inside_ of the subtree, so in the case of `C[ s1 ; s2 ]`, the gap between `s1` and `s2` is neither on the edge of the context hole (handled below) or properly inside of the `s1` or `s2` subtrees.  Therefore it would be invalidated.

Reply:
Yeah, that does seem like a natural way of generalizing this. Lots more code, though, Will keep in mind for a near-future PR.